### ISSUE TITLE
[Bug]: Sampling fails when temperature is 0

### ISSUE THREAD

**Issue by kogolobo** [turboderp/exllama#226]
This line in `generator.py` yields infinite logits when temperature is set to 0: https://github.com/turboderp/exllama/blob/c16cf49c3f19e887da31d671a713619c8626484e/generator.py#L106C1-L106C30
![image](https://github.com/turboderp/exllama/assets/44957968/ac3cb537-f766-4939-97a8-4fb48b6fa0ce)

Debugger result:
![image](https://github.com/turboderp/exllama/assets/44957968/6a0de06e-a4de-4e18-bea1-d7dddbfd4010)

**Comment by turboderp** [comment]
Temperature = 0 is an invalid argument the way temperature is defined here. I don't know if other implementations treat this as a special case or not, but the only sensible interpretation I can think of is that temperature = 0 should be equivalent to top-k = 1.

For the sake of numerical stability, a robust "fix" would have to take into account any numerical instability as the temperature *approaches* zero. What's the desired behavior here? Should any temperature less than some small threshold result just trigger greedy sampling?

**Comment by kogolobo** [comment]
I believe temp=0 should be equal to greedy decoding (as you mentioned same as top_k = 1). I really like your suggestion of selecting a small threshold that takes numerical instability into account😄

**Comment by turboderp** [comment]
But what's the typical behavior in other implementations? If I'm overriding undefined behavior arbitrarily anyway, I'd want to be as unsurprising as possible.

**Comment by kogolobo** [comment]
Good question. It seems that [HF is able to decide on the greedy decoding by other means](https://github.com/huggingface/transformers/blob/a6e6b1c622d8d08e2510a82cb6266d7b654f1cbf/src/transformers/generation/utils.py#L849), and does not even look at temperature setting. VLLM, however, similar to your suggestion, c[ompares temperature to a small constant and replaces it with "1.0"](https://github.com/vllm-project/vllm/blob/79af7e96a0e2fc9f340d1939192122c3ae38ff17/vllm/model_executor/layers/sampler.py#L192C1-L192C1) if it is below the constant.
